fog/ocall_oram_storage/trusted/src/lib.rs

```diff
@@ -571,3 +596,775 @@ extern "C" {
+            assert_eq!(data_scratch[0], a64_bytes(1));
+            assert_eq!(data_scratch[1], a64_bytes(1));
+            for data in &data_scratch[2..] {
+                assert_eq!(data, &a64_bytes(2));
+            }
+
+            // the first two meta should be 1's
+            assert_eq!(meta_scratch[0], a8_bytes(1));
+            assert_eq!(meta_scratch[1], a8_bytes(1));
+            for meta in &meta_scratch[2..] {
+                assert_eq!(meta, &a8_bytes(2));
+            }
+
+            st.checkin(131072 - 1, &mut data_scratch, &mut meta_scratch);
+        }
+
+        // Write 3's along branch 131072 / 2 + 1, and check if 1's and 2's are visible
+        {
+            st.checkout(131072 / 2 + 1, &mut data_scratch, &mut meta_scratch);
+
+            assert_eq!(data_scratch[16], a64_bytes(2));
+            assert_eq!(meta_scratch[16], a8_bytes(2));
+            for meta in &meta_scratch[0..16] {
+                assert_eq!(meta, &a8_bytes(0));
+            }
+
+            // write 3's
+            for data in data_scratch.iter_mut() {
+                *data = a64_bytes(3);
+            }
+            for meta in meta_scratch.iter_mut() {
+                *meta = a8_bytes(3);
+            }
+
+            st.checkin(131072 / 2 + 1, &mut data_scratch, &mut meta_scratch);
+        }
+
+        // Check that 3's are along branch at 131072/2 + 1
+        {
+            st.checkout(131072 / 2 + 1, &mut data_scratch, &mut meta_scratch);
+
+            for data in data_scratch.iter() {
+                assert_eq!(data, &a64_bytes(3));
+            }
+            for meta in meta_scratch.iter() {
+                assert_eq!(meta, &a8_bytes(3));
+            }
+
+            st.checkin(131072 / 2 + 1, &mut data_scratch, &mut meta_scratch);
+        }
+
+        // Check that 1's, 2's and 3's are visible along branch 131072 - 1
+        {
+            st.checkout(131072 - 1, &mut data_scratch, &mut meta_scratch);
+
+            // the first two data should be 1's
+            assert_eq!(data_scratch[0], a64_bytes(1));
+            assert_eq!(data_scratch[1], a64_bytes(1));
+            for data in &data_scratch[2..16] {
+                assert_eq!(data, &a64_bytes(2));
+            }
+            // this 3 at the root should be visible
+            assert_eq!(data_scratch[16], a64_bytes(3));
+
+            // the first two meta should be 1's
+            assert_eq!(meta_scratch[0], a8_bytes(1));
+            assert_eq!(meta_scratch[1], a8_bytes(1));
+            for meta in &meta_scratch[2..16] {
+                assert_eq!(meta, &a8_bytes(2));
+            }
+            // this 3 at the root should be visible
+            assert_eq!(meta_scratch[16], a8_bytes(3));
+
+            st.checkin(131072 - 1, &mut data_scratch, &mut meta_scratch);
+        }
+    }
+
+    // Test what happens when we start screwing with the untrusted memory,
+    // by hammering it with junk value
+    #[test]
+    fn exercise_oram_storage_hammer_data() {
```

Comment:
It's really odd to me that this isn't failing on line 314. I must be missing something about the way we calculate our hashes. If we go through and hammer the data, shouldn't the newly calculated hash be different throughout the tree and not just at the merkle root?

Reply:
On further investigation. This is because for some reason, hammer meta data is getting that the self.meta_scratch_buffer[idx] == Default::default() is false, whereas hammer data is having this condition return true, which seems to be what's causing the last hash to be coming out as 0s.